.github/workflows/deploy-dotorg.yml

```diff
@@ -7,17 +7,33 @@ on:
 
 jobs:
   release:
-    name: New Release
+    name: Performance Lab plugin deploy
     runs-on: ubuntu-latest
     steps:
-      - name: Checkout code
+      - name: Checkout
         uses: actions/checkout@v3
+
+      - name: Setup Node.js (via .nvmrc)
+        uses: actions/setup-node@v3
+        with:
+          node-version-file: '.nvmrc'
+          cache: npm
+
+      - name: Install dependencies
+        run: npm ci
+
+      - name: Build Performance Lab plugin
+        run: npm run build:plugin:performance-lab
+
       - name: WordPress plugin deploy
         uses: 10up/action-wordpress-plugin-deploy@stable
         env:
           SVN_PASSWORD: ${{ secrets.SVN_PASSWORD }}
           SVN_USERNAME: ${{ secrets.SVN_USERNAME }}
+          BUILD_DIR: 'build/performance-lab'
+          ASSETS_DIR: '.wordpress-org'
           SLUG: performance-lab
```

Comment:
What is the purpose of all these changes? I don't see a good reason for them. The main Performance Lab process has not needed a build process so far, and I don't see why it would need one now.

I understand we need to fix the process for standalone plugins, but for Performance Lab it has worked perfectly well the way it has been so far. The additions here, if anything, will make the process slower as far as I can tell, due to installing NPM dependencies etc.

Reply:
@felixarntz Thanks for pointing this out. I have updated the PR description with the required details on why we need to switch to another build setup.

Now as you mentioned:

>.gitattributes has the benefit that it is also processed by GitHub. When you click "Download ZIP", you only get a ZIP of the actual Performance Lab plugin without the dev ops files. With .distignore that is no longer the case.

In cases where the PL plugin has such files that are not in the git tree and are generated by a build setup, this thing will not be the ideal solution. IMO prod-ready zip should be tagged to plugin release assets which also provide the flexibility to download releases with GitHub APIs.

Now a build step for the main plugin brings a few things to the table:
- A prod-ready zip is tagged on the release assets.
- In case someone needs the PL plugin built locally, it provides the way to do it.
- Since we already created a build system for standalone plugins, the same will also work with the main plugin. So I think it would be better to take the control over distribution creation process of the PL plugin as well from deploy action like we are doing for other plugins.

Reply:
@thelovekesh 
> In cases where the PL plugin has such files that are not in the git tree and are generated by a build setup, this thing will not be the ideal solution.

Can you provide a concrete example where that is the case? I don't see any files being generated by the build setup for Performance Lab.

> IMO prod-ready zip should be tagged to plugin release assets which also provide the flexibility to download releases with GitHub APIs.

That is a good point, but it doesn't require us to have a build step. This can be accomplished with the 10up deploy action alone, see https://github.com/10up/action-wordpress-plugin-deploy?tab=readme-ov-file#inputs.

I think the reasons you're pointing out are valid in theory, but not in practice. As far as I can tell, the Performance Lab plugin doesn't require to be built, it works just from the source - unless I'm missing something. Unless there is a concrete usage that requires the build process, I think we should avoid it, as it introduces a process for something that doesn't need it, which makes things slower and is one more thing to maintain.

Reply:
@felixarntz I think there are two possible scenarios here:
- PL plugin which doesn't have any build setup as of now, so works well with the current `.gitattributes` based setup.
- Standalone plugins, where:
    - ILO and Web Worker Offloading have a build step that includes vendor library.
    - We need to generate builds for these which should have the vendor files(which are git ignored) and should not have config/WP Assets files.

Now we have a webpack-based build step(by keeping standalone plugins in mind) that will build plugins and ignore files that are not required but include files that are git ignored(which doesn't seem possible with `.gitattributes`), hence we switched to `.distignore`.

> Unless there is a concrete usage that requires the build process, I think we should avoid it, as it introduces a process for something that doesn't need it, which makes things slower and is one more thing to maintain.

Now the same build setup can be used to generate for build PL plugin as well at no extra cost, I don't think it will require any maintenance.

Reply:
@thelovekesh Based on your reply, there's no reason to use a build process for Performance Lab. I completely agree the other standalone plugins need it, but that doesn't mean we need to use it for PL as well. Even if the maintenance cost is little, there is no benefit to using it for Performance Lab. So let's use it only for the standalone plugins.

> Now we have a webpack-based build step(by keeping standalone plugins in mind) that will build plugins and ignore files that are not required but include files that are git ignored(which doesn't seem possible with `.gitattributes`), hence we switched to `.distignore`.

I don't understand this. `.gitattributes` is used by GitHub e.g. to ignore files when users download a ZIP file, and it's also used by the 10up deploy script. I don't think it interferes with a custom Webpack build process. Of course 3rd party libraries like `web-vitals` need to be in `.gitignore`, which means they don't even need to be listed in `.gitattributes` because they won't be part of the repo. For the build process of the standalone plugins, that means they have to run `npm install` before building the standalone plugins, and those aren't subject to `.gitattributes` anyway - only the main PL plugin is.

In other words: I think as long as we continue to separate the processes, one for deploying PL, another one for building and deploying standalone plugins, there should not be an issue. If I'm missing something, can you please provide more detail on which problem you're encountering because of `.gitattributes`?

Reply:
> In other words: I think as long as we continue to separate the processes, one for deploying PL, and another one for building and deploying standalone plugins, there should not be an issue. If I'm missing something, can you please provide more detail on which problem you're encountering because of .gitattributes?

Okay, I see. In essence, my goal was to make this procedure more efficient for every plugin. I believe that if we leave `.gitattributes` alone, I will have to explicitly specify which patterns to ignore for the standalone plugin build rather than obtaining them from a global `.distignore`.

I don't mind if the release workflow for PL and standalone plugins is kept separate, but I still think that the build setup should retain `.distignore` since the 10up deploy action [supports](https://github.com/10up/action-wordpress-plugin-deploy?tab=readme-ov-file#excluding-files-from-deployment) it (if you want to handle the build from the 10up action) or with plugin build setup if you agree to keep `.distignore`. 

Attaching the build zip to the release section is always an option if we are only keeping the `.gitattributes` because it offers processed ZIP. In any case, once we have build files for the PL plugin, we will need to delete it. Who knows, maybe we'll use WordPress components for https://github.com/WordPress/performance/issues/1032. However, as you pointed out, that's only a theory 😎

Reply:
> I believe that if we leave `.gitattributes` alone, I will have to explicitly specify which patterns to ignore for the standalone plugin build rather than obtaining them from a global `.distignore`.

This is the part I do not understand. How does this work when using `.distignore` but not when using `.gitattributes`? As far as my understanding goes, neither of these two files are used by the 10up deploy action when we specify a custom `BUILD_DIR`, as you pointed out before. So what's the difference between these two files for the build process you're adding here? Where in the build process is `.distignore` used? Maybe I'm missing something, so can you please clarify?

Reply:
Aah okay. I am using `.distignore` to pass patterns to be ignored in the `copy-webpack-plugin`.

https://github.com/WordPress/performance/blob/9f38e400ecdac6ddf3abc505adf0d44e11db273b/webpack.config.js#L82

Reply:
Yes @felixarntz, it makes sense to me. I believe we should keep the `.gitattributes` file as it is also processed by https://github.com/github-linguist/linguist (and the AMP plugin uses it like https://github.com/ampproject/amp-wp/blob/develop/.gitattributes). This is after researching the usage of `.gitattributes` in a GitHub repository in greater detail.

The only drawback is that the 10up deploy action will prioritize `.distignore` if we retain both `.gitattributes` and `.distignore`([ref](https://github.com/10up/action-wordpress-plugin-deploy/blob/abb939a0d0bfd01063e8d1933833209201557381/deploy.sh#L94)). 

Since Webpack is now using `.distignore`, I would suggest moving it to `bin/webpack`. As a result, the standalone plugin will continue to use `.distignore`, and the PL plugin will continue to use `.gitattributes`. 

We can easily transfer it to the plugin root if we ever decide to use `.distignore` for PL plugins as well.

Reply:
Also, I don't see any issues if we use `.distignore` with the 10up deploy action as well. Then it will be like:
- Use `.distignore` for creating distribution builds.
- Use `.gitattributes` for GitHub related chores and ZIP processing.

Reply:
@thelovekesh 
> Since Webpack is now using `.distignore`, I would suggest moving it to `bin/webpack`. As a result, the standalone plugin will continue to use `.distignore`, and the PL plugin will continue to use `.gitattributes`. 

This sounds great to me.

> Also, I don't see any issues if we use `.distignore` with the 10up deploy action as well. Then it will be like:
> - Use `.distignore` for creating distribution builds.
> - Use `.gitattributes` for GitHub related chores and ZIP processing.

While we could do that, this would mean we would still need to basically maintain the large file list in two places. So I would prefer to go with your prior suggestion and use `.gitattributes` in the root (for both GitHub and Performance Lab build/deployment) and `.distignore` in `bin/webpack` (for the standalone plugins build/deployment).

Reply:
@felixarntz It looks like the ignore file is not working as expected when moving it to `bin/webpack` directory since it's adding cwd path to patterns. Hence I have directally added the ingore paths in Webpack config for now since they are very few. Please let me know if it looks good to you.

https://github.com/WordPress/performance/blob/41b45ca5826671fe5cb7599704a688310336d0d3/webpack.config.js#L68-L71

Reply:
@thelovekesh Thanks, that works for me! Given it's so few files, it should be okay.